diagnostics/diagnostics_manager.rs

```diff
@@ -0,0 +1,217 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use std::{
+    collections::HashSet,
+    error::Error,
+    fmt::format,
+    future::Future,
+    net::SocketAddr,
+    path::PathBuf,
+    pin::Pin,
+    sync::{Arc, Mutex},
+    time::Instant,
+};
+
+use error::TypeDBError;
+use resource::constants::diagnostics::REPORTING_URI;
+use tonic::Status;
+use tonic_types::StatusExt;
+
+use crate::{
+    metrics::{ActionKind, DatabaseMetrics, LoadKind},
+    monitoring_server::MonitoringServer,
+    reporter::Reporter,
+    Diagnostics,
+};
+
+macro_rules! diagnostics_method {
+    ($(
+        pub fn $fn_name:ident(&self, $($arg_name:ident: $arg_type:ty),* $(,)?) ;
+    )*) => {
+        $(
+            pub fn $fn_name(&self, $($arg_name: $arg_type),*) {
+                self.diagnostics.$fn_name($($arg_name),*)
+            }
+        )*
+    };
+}
+
+#[derive(Debug)]
+pub struct DiagnosticsManager {
+    diagnostics: Arc<Diagnostics>,
+    reporter: Reporter,
+    monitoring_server: MonitoringServer,
+}
+
+impl DiagnosticsManager {
+    pub fn new(
+        deployment_id: String,
+        server_id: String,
+        distribution: String,
+        version: String,
+        data_directory: PathBuf,
+        monitoring_port: u16,
+        is_monitoring_enabled: bool,
+        is_reporting_enabled: bool,
+    ) -> Self {
+        let diagnostics = Arc::new(Diagnostics::new(
+            deployment_id.clone(),
+            server_id,
+            distribution,
+            version,
+            data_directory.clone(),
+            is_reporting_enabled,
+        ));
+        let reporter =
+            Reporter::new(deployment_id, diagnostics.clone(), REPORTING_URI, data_directory, is_reporting_enabled);
+        let monitoring_server = MonitoringServer::new(diagnostics.clone(), monitoring_port, is_monitoring_enabled);
+
+        Self { diagnostics, reporter, monitoring_server }
+    }
+
+    diagnostics_method! {
+        pub fn submit_database_metrics(&self, database_metrics: HashSet<DatabaseMetrics>);
+        pub fn submit_error(&self, database_name: Option<&str>, error_code: String);
+        pub fn submit_action_success(&self, database_name: Option<&str>, action_kind: ActionKind);
+        pub fn submit_action_fail(&self, database_name: Option<&str>, action_kind: ActionKind);
+        pub fn increment_load_count(&self, database_name: &str, connection_: LoadKind);
+        pub fn decrement_load_count(&self, database_name: &str, connection_: LoadKind);
+    }
```

Comment:
Why are both `DiagnosticsManager` and `Diagnostics` responsible for this? As far as I can tell, these are only called from a handful of places.

`DiagnosticsResultTracker`, e.g., doesn't need anything manager-specific, and should just have a reference to `Diagnostics` instead. Same goes for `synchronize_database_metrics()`. Then this macro can go away.

Reply:
It was simpler for me to think about resource sharing in concurrency this way. `Diagnostics` are initialized by the `DiagnsoticsManager` that also controls all the `MonitoringServer`s and `Reporter`s, so it can share the ownership once on the setup. Then, if you need to work with `Diagnostics`, you need to do it through the `DiagnosticsManager`, which can, if needed, control the concurrency issues if needed. 

It's not used now as this control is moved to `Diagnostics`, and I don't have the plan to move it higher. 

In case of actions and errors, right now it can really be removed from the manager as the `ResultTracker` covers these needs. However, I still need `submit_database_metrics`, `increment_load_count`, and `decrement_load_count` which are used against the manager, and I'd like to encapsulate it instead of exposing the `Diagnostics` out of the `Manager`.

Is it worth it?

Reply:
UPD. As I am going to remove the tracker, I think that it is worth preserving the intermediate encapsulating calls of `DiagnosticsManager`

Reply:
I was originally thinking of exposing the diagnostics field through a method and using that when needed. There's no danger in exposing it, and it's no less clear at the call site (arguably, more clear that `submit_` methods only affect the diagnostics registry and not any services). In any case, the encapsulation is kinda meaningless here since all the same behaviour is available on the manager anyway, and no extra checking is done.

But now I'm thinking, do we even need a `DiagnosticsManager`? It's not managing anything, it just groups the `Diagnostics` with a couple of independent services.

Reply:
I do think that it's easier to have this (or any other) intermediate layer, because `Diagnostics` are aware of `Reporter`s and `MonitoringServer`s (they have special formatting for them), and the `Reporter` and the `MonitoringServer` are aware of `Diagnostics`. So this layer of managing all these connections makes it simpler, but I may be too used to this structure, thus hesitant to change anything. 

I don't feel a strong urge to deconstruct it at the moment, and I don't like the diagnostics getter... Yet.